feat: add SystemTap

ok-ryoko opened the pull request:
[SystemTap] is a Red Hat-driven Linux instrumentation framework that’s been around since 2005. It exposes a scripting language inspired by C and awk. I’ve written a complete [grammar] and have shown up to contribute nvim-treesitter queries.

Here’s the highlights query in action on some of my code:

![systemtap-nvim-treesitter-highlights](https://github.com/nvim-treesitter/nvim-treesitter/assets/59705845/654c418d-ef66-426c-b841-2ea1684875c4)

There’s very little SystemTap in the wild, so I would consider this PR to be a relatively low priority.

Your feedback is greatly appreciated, even if this PR doesn’t succeed. Thank you for your work on nvim-treesitter!

[grammar]: https://github.com/ok-ryoko/tree-sitter-systemtap
[SystemTap]: https://sourceware.org/systemtap/

ObserverOfTime commented on queries/systemtap/highlights.scm:
```diff
@@ -0,0 +1,156 @@
+(identifier) @variable
+
+(preprocessor_macro_definition
+  name: (identifier) @define)
```
```suggestion
  name: (identifier) @function.macro)
```

ObserverOfTime commented on queries/systemtap/highlights.scm:
```diff
@@ -0,0 +1,156 @@
+
+(preprocessor_macro_expansion) @function.macro
+
+(preprocessor_constant) @constant.macro
```
```suggestion
(preprocessor_constant) @constant.builtin
```

ObserverOfTime commented on queries/systemtap/highlights.scm:
```diff
@@ -0,0 +1,156 @@
+
+(type) @type.builtin
+
+(aggregation_operator) @function.macro
```
```suggestion
(aggregation_operator) @attribute
```

ObserverOfTime commented on queries/systemtap/highlights.scm:
```diff
@@ -0,0 +1,156 @@
+(call_expression
+  function: (identifier) @function.call)
+
+(
+  (call_expression
```
Formatting.

```suggestion
((call_expression
```

ObserverOfTime commented on queries/systemtap/highlights.scm:
```diff
@@ -0,0 +1,156 @@
+    function: (identifier) @function.builtin)
+  (#any-of? @function.builtin
+    "print" "printd" "printdln" "printf" "println"
+    "sprint" "sprintd" "sprintdln" "sprintf" "sprintln")
+)
```
```suggestion
    "sprint" "sprintd" "sprintdln" "sprintf" "sprintln"))
```

ObserverOfTime commented on queries/systemtap/highlights.scm:
```diff
@@ -0,0 +1,156 @@
+    "sprint" "sprintd" "sprintdln" "sprintf" "sprintln")
+)
+
+(
+  (identifier) @variable.builtin
+  (#lua-match? @variable.builtin "^\$+[0-9A-Z_a-z]+\$*$")
+)
```
```suggestion
((identifier) @variable.builtin
  (#lua-match? @variable.builtin "^\$+[0-9A-Z_a-z]+\$*$"))
```

ObserverOfTime commented on queries/systemtap/highlights.scm:
```diff
@@ -0,0 +1,156 @@
+] @exception
+
+[
+  "%("
+  "%)"
+  "%:"
+  "%?"
+  "%{"
+  "%}"
```
Why not `@punctuation.bracket`?

ok-ryoko replied:
The tokens `%(`, `%)`, `%:` and `%?` are meaningful only to the preprocessor, being interpreted in a separate pass before the SystemTap script gets parsed.

The delimiters `%{` and `%}` are used only for embedded C code, which is transcribed verbatim.

Another reason for choosing distinct captures for these tokens is to achieve consistency with the behavior of the language server.

ObserverOfTime reviewed (changes requested):
I don't quite understand what these nodes represent:
```
(script_argument_string)
(script_argument_number)
(preprocessor_tokens)
(embedded_code)
```

amaanq commented on queries/systemtap/highlights.scm:
```diff
@@ -0,0 +1,156 @@
+  "break"
+  "continue"
+  "delete"
+  "global"
```
Could be `@storageclass`

amaanq commented on queries/systemtap/highlights.scm:
```diff
@@ -0,0 +1,156 @@
+] @punctuation.bracket
+
+[
+  "break"
+  "continue"
```
Should be `@repeat`

amaanq commented on queries/systemtap/highlights.scm:
```diff
@@ -0,0 +1,156 @@
+] @punctuation.bracket
+
+[
+  "break"
```
Should be `@conditional` or `@repeat`

amaanq commented on queries/systemtap/highlights.scm:
```diff
@@ -0,0 +1,156 @@
+  "global"
+  "limit"
+  "next"
+  "private"
```
Could be `@type.qualifier`

ok-ryoko replied:
I’m wary of using `@type` here since `private` can be used only for variable declarations and function definitions. Moreover, SystemTap has no user-defined types.

amaanq replied:
private is a qualifier though, isn't it? `@type.qualifier` really works for any sort of qualifier

ok-ryoko replied:
Is it the convention in nvim-treesitter to be able to use `@type.qualifier` to qualify entities other than types? If so, then I will update the capture.

ObserverOfTime replied:
One day it will be renamed to something more sensible…

amaanq commented on queries/systemtap/highlights.scm:
```diff
@@ -0,0 +1,156 @@
+  "%?"
+  "%{"
+  "%}"
+  (preprocessor_tokens)
```
Maybe this can be injected w/ itself? If it's anything like Rust/C

ok-ryoko replied:
I have addressed this in [my reply](https://github.com/nvim-treesitter/nvim-treesitter/pull/5097#issuecomment-1634546706) to the main conversation. Could you please take a look there?

amaanq commented:
Also you need an entry in `lockfile.json` w/ the latest revision @ok-ryoko

ok-ryoko commented:
Would you also like me to set definition scopes in the locals query? Would that constitute a meaningful improvement to the submission quality?

ObserverOfTime commented:
> script_argument_string and script_argument_number refer to the special identifiers @1, …, @99 and $1, …, $99, respectively. These represent indexed command-line arguments to the script. For example, if your script accepts an integer as the first command-line argument, then you could access that integer using $1.

How about `@parameter` then?

>  Does nvim-treesitter expose a quick toggle for injections?

We have a toggle for highlights but not injections. Maybe we should. :thinking:

ok-ryoko commented:
In review of my comments, I have not made a compelling case against injecting C highlights. Having highlights for embedded C would help differentiate embedded blocks from preprocessor blocks. Moreover, the location of embedded C in SystemTap scripts is well-defined, whereas preprocessor blocks can appear virtually anywhere (although the grammar doesn’t allow that). Thus, I added an injections capture and switched to `@punctuation.bracket` for `%{` and `%}`.

> How about `@parameter` then?

My impression of the `@parameter` capture is that it is reserved for declarations, as in the C highlights query. If we used `@parameter` to color the script arguments in the script body, then I would expect to have to use `@parameter` to color every function’s parameters in the corresponding function body, locals notwithstanding. From my reading of nvim-treesitter queries, this doesn’t seem to be a common pattern.
